### ISSUE TITLE
[java] PMD must display number of rules contravened or errors found. 

### ISSUE THREAD

**Issue by Fernal73** [pmd/pmd#1995]
**Affects PMD Version:** 
6.17

**Rule:**
All rulesets. 

**Description:**
PMD output does not inform the user as to the number of rules contravened while running the tool. The user has to look at the output file. 

**Code Sample demonstrating the issue:**

```
Sep 01, 2019 9:42:45 AM net.sourceforge.pmd.cache.FileAnalysisCache loadFromFile
INFO: Analysis cache loaded
Sep 01, 2019 9:43:02 AM net.sourceforge.pmd.cache.AbstractAnalysisCache checkValidity
INFO: Analysis cache invalidated, execution classpath changed.
Sep 01, 2019 9:43:07 AM net.sourceforge.pmd.cache.FileAnalysisCache persist
INFO: Analysis cache updated
```

**Running PMD through:**
[CLI | Ant]

**Comment by ThanasisGkanos** [comment]
Looking into this

**Comment by Fernal73** [comment]
https://pmd.github.io/latest/pmd_userdocs_cli_reference.html#exit-status

Would it be possible to return a -ve exit code instead of 1 for "Couldn't understand command-line parameters or PMD exited with an exception" and the number of violations encountered as the exit code when violations are detected? Everything else can remain the same.

**Comment by Fernal73** [comment]
https://pmd.github.io/latest/pmd_userdocs_tools_ant.html#parameters

The ant task allows capture of the number of violations via failuresPropertyName. This will enable capturing the same via CLI.

**Comment by adangel** [comment]
> Would it be possible to return a -ve exit code instead of 1 for "Couldn't understand command-line parameters or PMD exited with an exception" and the number of violations encountered as the exit code when violations are detected? Everything else can remain the same.

Do you mean, to return a negative exit code? I don't know the "-ve exit code"...
As you see in the docs, we currently have 3 exit codes, which means, that "0", "1" and "4" are taken. Whatever we return, I think, the 3 cases still need to be distinguishable. e.g. if the build fails, because PMD didn't run at all (1) or because there are rule violations (4) is an important difference, that helps in fixing the problem and looking at the right place (pmd configuration, build scripts, vs. source code).

At least in PMD 6 we need to stay compatible, so we can only use exit codes above 4.... Or do you mean with "-ve" adding a new commandline option, called "--violations-as-exit-code" (or short -ve)? That would indeed be possible. But I would still like to understand your use case.

Currently:
* PMD exits with 0: Everything is fine, nothing to do
* PMD exits with 1: PMD didn't run, need to look at the build script
* PMD exits with 4: violations found, need to look at the report

With "-ve":
* PMD exits with 0: Everything is fine, nothing to do
* PMD exits with e.g. 3: there are 3 violations found, need to look at the report
* PMD exits with 65535: PMD didn't run, need to look at the build script (or there are really 65535 violations)

**Comment by Fernal73** [comment]
I'm fine with 65535.
That's an outlier if it happens.
The possibility is remote.

https://softwareengineering.stackexchange.com/questions/170694/why-error-codes-are-negated/170707

**Comment by Fernal73** [comment]
With a -ve error code, there'd be no ambiguity. It'd be nice to capture the
number of violations without parsing the text output. That's my usecase.

**Comment by Fernal73** [comment]
https://stackoverflow.com/a/13247335/3924108

**Comment by Fernal73** [comment]
I understand this might impact programs or scripts that depend on these error codes. Would moving it to 7.0 provide adequate advance warning?

**Comment by Fernal73** [comment]
I'm inclined to switch the 1 error code to -1 with all future error codes as negative values. 
Additionally, a message to be displayed in all 6.xx versions prior to 7.00 that the error codes will change to different values in version 7.0 and scripts and programs that depend on existing values will have to be modified. The positive number returned will always be the number of violations encountered unless overridden to return 0.

**Comment by oowekyala** [comment]
Quoting from https://www.shellscript.sh/exitcodes.html

> Exit codes are a number between 0 and 255 [...]. Other numbers can be used, but these are treated modulo 256, so exit -10 is equivalent to exit 246, and exit 257 is equivalent to exit 1. 

Testing this in my shell:
```
16:22 @~$ bash -c "exit 258;"
16:22 @~$ echo $?
2
16:22 @~$ bash -c "exit -1;"
16:23 @~$ echo $?
255
```

I don't think returning the number of violations as an exit code is an option. That's anyway not what the return code is meant for. 

@Fernal73 I don't really understand your use case though. What language/shell do you use to start PMD? Do you parse the report output? Do you expect to be able to change the renderer from outside your script/program? If this last question is no, then you could define your custom renderer.

**Comment by Fernal73** [comment]
No intention to parse the output, as yet.
The original request was to display the number of errors.
My mistake about the shell error codes. A glaring one.

**Comment by ThanasisGkanos** [comment]
Hey guys sorry for taking so long to come back on this. I can see a similar implementation that already exists [here](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/renderers/TextColorRenderer.java#L113). This is triggered by passing the flag: `-f textcolor `. What exactly are we looking from this enhancement?

**Comment by oowekyala** [comment]
@ThanasisGkanos Tbh even with *while running the tool* I think the question is very vague. If it was the case that OP's looking for a progress indicator, then changing the exit code would have been pointless. 

@Fernal73 *Why* do you want to read the number of rules? Do you need to parse it or do you just think it would be useful info for an end-user? Are you asking for a progress indicator during the run, or just a summary line at the end? What renderer are you using?

**Comment by Fernal73** [comment]
The renderer is text. It's needed so that the file does not have to be
looked at to know if errors are present or not. There is no other indicator
in the logged output presently.

**Comment by Fernal73** [comment]
With format textcolor, the output is logged to the file , not to the
console. The file still needs to be looked at.
There ought to be an error count logged to console , irrespective.

Secondly, there are couple of problems with the error code returned
currently. If failOnViolations is set and PMDException occurs, the error
code returned is 0, not 1.
 If violations are present as well, the error code returned is 4, not 1. Is
this the norm or expected behaviour?

**Comment by Fernal73** [comment]
@oowekyala

Summary line with separator for each type including exceptions or errors as defined by ret code 1 is fine.

**Comment by Fernal73** [comment]
https://github.com/pmd/pmd/issues/2006

Linking above issue in for further context.

**Comment by Fernal73** [comment]
https://softwareengineering.stackexchange.com/questions/153926/non-zero-exit-status-for-clean-exit

I'll add this post as to how error codes are returned. I'm inclined to go with negative or high positive numbers for errors. However, I don't think PMD needs to change anything about the existing error codes unless you're looking for more granularity in the future.

**Comment by adangel** [comment]
As regarding to this issue, this is only about *displaying* the number of rule violations and *not* changing the exit code. Let's not scope creep this issue.
If you still think, the exit code should encode this information and you can provide a useful use case, please create a separate issue for that, thanks.

**Comment by Fernal73** [comment]
Do we need a threshold limit (for number of violations)  at which PMD
terminates early?
That will be another issue.

**Comment by Fernal73** [comment]
https://metacpan.org/pod/perlcritic#OPTIONS

I've been looking at PerlCritic for my perl programs.

We can cannibalise command line options from this static code analyser.

At the outset, 

--top [ N ],
--statistics-only,
And
--count
-C

appear to be good candidates to implement.

**Comment by matheuslcruz** [comment]
> If failOnViolations is set and PMDException occurs, the error code returned is 0, not 1. That's not expected.

Any news about this? This unexpected behavior still happening in v6.28.0

**Comment by Fernal73** [comment]
> > If failOnViolations is set and PMDException occurs, the error code returned is 0, not 1. That's not expected.
> 
> Any news about this? This unexpected behavior still happening in v6.28.0

Separate issue needed?

**Comment by adangel** [comment]
@Fernal73 Yes, please create a separate issue for exit codes. This is independent of your feature request. I'll hide the unrelated posts.

Back to your feature request: Do I understand it correctly, that you want something like [printFailingErrors](https://maven.apache.org/plugins/maven-pmd-plugin/check-mojo.html#printFailingErrors). That's something very handy that the maven-pmd-plugin has: In case the build is failing because of violations, these violations are printed on the console (the build log), so that you don't need to open the report pmd.xml file. You directly see, that it failed and why and can start fixing issues.

**Comment by Fernal73** [comment]
Ok. I will do that.
Is it safe to assume that there's no need to test that the exit code is
still zero when an exception occurs? Of course, it is. Why else would it be
brought to your attention once more?

**Comment by Fernal73** [comment]
> 
> Back to your feature request: Do I understand it correctly, that you want something like [printFailingErrors](https://maven.apache.org/plugins/maven-pmd-plugin/check-mojo.html#printFailingErrors). That's something very handy that the maven-pmd-plugin has: In case the build is failing because of violations, these violations are printed on the console (the build log), so that you don't need to open the report pmd.xml file. You directly see, that it failed and why and can start fixing issues.

Best to be consistent across build methods.

**Comment by adangel** [comment]
Ok, I went through all the comments and hid all that were off topic.
The task here is as follows:

When you run PMD on the command line (see https://pmd.github.io/latest/pmd_userdocs_cli_reference.html), PMD just executes and terminates when it has finished. There is no visual output at all to tell the user e.g. "Found 11 violations. And there were 2 processing errors.".
Or in case, no problems have been detected, the output could be something like "Found no violations.".

See also https://github.com/pmd/pmd/wiki/Newcomers%27-Guide

**Comment by abhinavashok** [comment]
Is anyone working on this ?

**Comment by adangel** [comment]
@abhinavashok I don't think so.

**Comment by shravanatirtha** [comment]
Hi, I would want to work on this issue.

**Comment by DragonFSKY** [comment]
I don't see an active PR or assignee for this. I'm preparing a small CLI-only implementation that prints the existing violation/processing-error counts through the CLI reporter without changing report output or exit codes. Happy to defer if someone is actively working on it.
